configure.ac: Fix autoreconf.

orbea opened the pull request:
The goal behind this PR is to fix `autoreconf -fi`.
```
$ autoreconf -fi
aclocal: warning: couldn't open directory 'm4': No such file or directory
autoreconf: configure.ac: AM_GNU_GETTEXT is used, but not AM_GNU_GETTEXT_VERSION
libtoolize: putting auxiliary files in '.'.
libtoolize: copying file './ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
configure.ac:10: installing './compile'
configure.ac:11: installing './config.guess'
configure.ac:20: error: required file './config.rpath' not found
configure.ac:11: installing './config.sub'
configure.ac:6: installing './install-sh'
configure.ac:6: installing './missing'
Makefile.am: installing './INSTALL'
Makefile.am: installing './depcomp'
autoreconf: automake failed with exit status: 1
```
While `./autogen.sh` works, it can be nice to use the `autoreconf` tool instead.

So my understanding based on some searching online is that using only `AM_GNU_GETTEXT` will fail to copy the required files to the build directory. This can be solved by adding `AM_GNU_GETTEXT_VERSION`, I set this to the version `0.18.2` because this is the oldest version that avoided the following warnings which also currently occur with `autogen.sh`.
```
configure.ac:20: warning: The 'AM_PROG_MKDIR_P' macro is deprecated, and its use is discouraged.
configure.ac:20: You should use the Autoconf-provided 'AC_PROG_MKDIR_P' macro instead,
configure.ac:20: and use '$(MKDIR_P)' instead of '$(mkdir_p)'in your Makefile.am files.
```
This is a gettext bug which has been solved in `0.18.2`, but newer versions will also work for me and the version can be changed of course. Repology can be used to get an idea of what versions of `gettext` distributions are using.

https://repology.org/metapackage/gettext/versions

However while adding `AM_GNU_GETTEXT_VERSION([0.18.2])` fixes `autoreconf` without warnings it exposes a new issue for `./confgure`.
```
config.status: error: po/Makefile.in.in was not created by intltoolize.
```
This appears to be a problem with using both gettext and intltool which are not compatible. Please see these links for some documentation.

https://github.com/amsynth/amsynth/issues/108
https://github.com/amsynth/amsynth/commit/5794c69e17b7e6ecada341687eaf10f0920eb810
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=724555
https://github.com/GNOME/gnome-builder/commit/0c9e17836a43ce7e894cb865c22ab8e9a27db1e5

I first tried to remove gettext, but suffice to say this was a bad idea and did not work out so well... SO next removed `IT_PROG_INTLTOOL` and added `AC_DEFINE_UNQUOTED(GETTEXT_PACKAGE,"$GETTEXT_PACKAGE", [Gettext package])` which I noticed various GNOME projects added for similar changes. I am not sure this is required?

Additionally based on various GNOME projects I looked at it seemed it might be good to also add `AC_DEFINE_UNQUOTED(GETTEXT_PACKAGE,"$GETTEXT_PACKAGE", [Gettext package])`, but I am not sure this is required.

Finally both `autoreconf` and `configure` worked, but now `make` failed because it was missing `po/Makevars`. My understanding is that it should be copied from `po/Makevars.template` and then modified if required. The only change I required was adding `--from-code=UTF-8` since some files failed to compile without it.

Some more documentation.

https://www.guyrutenberg.com/2014/11/01/gettext-with-autotools-tutorial/
https://www.gnu.org/software/gettext/manual/html_node/po_002fMakevars.html

With this everything seems to work, but it also seemed to make sense to remove some no longer needed intl files from `Makefile.am` and the no longer needed `intltoolize` line from `autogen.sh`. Lastly I added a second commit that removes these files and the `intl/` directory from the repo as both `autogen.sh` and `autoreconf` seem capable of copying the required files. This commit can be dropped or altered if desired.

orbea commented:
Please review this, while this works for me I am not sure if I overlooked or failed to understand something.

Alexey-Yakovenko commented:
There's a way to test that everything is still working correctly, by running `make dist-check`.
This would generate a tarball, then extract it to a temp folder, and build the whole app from it.
Then it run `make install` and `make uninstall` and verify that they are working too.
Hope this helps. I won't be able to test this until next week.

orbea commented:
I am not seeing a `dist-check` target, but I'll assume you meant `distcheck`?

Unfortunately this fails in master for me.
```
Making check in po
make[2]: Entering directory '/tmp/deadbeef/deadbeef-devel/_build/sub/po'
make[2]: *** No rule to make target '../../../translation/plugins.c', needed by 'deadbeef.pot'.  Stop.
make[2]: Leaving directory '/tmp/deadbeef/deadbeef-devel/_build/sub/po'
make[1]: *** [Makefile:906: check-recursive] Error 1
make[1]: Leaving directory '/tmp/deadbeef/deadbeef-devel/_build/sub'
make: *** [Makefile:1119: distcheck] Error 1
```
And my PR fails elsewhere, so I suspect I did miss something...
```
gcc -c -DLOCALEDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/share/locale\" -DLOCALE_ALIAS_PATH=\"/tmp/deadbeef/deadbeef-devel/_inst/share/locale\" -DLIBDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DBUILDING_LIBINTL -DBUILDING_DLL -DIN_LIBINTL -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DINSTALLDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DNO_XMALLOC -Dset_relocation_prefix=libintl_set_relocation_prefix -Drelocate=libintl_relocate -DDEPENDS_ON_LIBICONV=1 -DHAVE_CONFIG_H -I. -I../../../intl -I..  -g -O2  -D_GNU_SOURCE  -DLIBDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DPREFIX=\"/tmp/deadbeef/deadbeef-devel/_inst\" -DDOCDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/share/doc/\" -DDDB_WARN_DEPRECATED=1 -fvisibility=hidden  ../../../intl/lock.c
gcc -c -DLOCALEDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/share/locale\" -DLOCALE_ALIAS_PATH=\"/tmp/deadbeef/deadbeef-devel/_inst/share/locale\" -DLIBDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DBUILDING_LIBINTL -DBUILDING_DLL -DIN_LIBINTL -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DINSTALLDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DNO_XMALLOC -Dset_relocation_prefix=libintl_set_relocation_prefix -Drelocate=libintl_relocate -DDEPENDS_ON_LIBICONV=1 -DHAVE_CONFIG_H -I. -I../../../intl -I..  -g -O2  -D_GNU_SOURCE  -DLIBDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DPREFIX=\"/tmp/deadbeef/deadbeef-devel/_inst\" -DDOCDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/share/doc/\" -DDDB_WARN_DEPRECATED=1 -fvisibility=hidden  ../../../intl/relocatable.c
gcc -c -DLOCALEDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/share/locale\" -DLOCALE_ALIAS_PATH=\"/tmp/deadbeef/deadbeef-devel/_inst/share/locale\" -DLIBDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DBUILDING_LIBINTL -DBUILDING_DLL -DIN_LIBINTL -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DINSTALLDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DNO_XMALLOC -Dset_relocation_prefix=libintl_set_relocation_prefix -Drelocate=libintl_relocate -DDEPENDS_ON_LIBICONV=1 -DHAVE_CONFIG_H -I. -I../../../intl -I..  -g -O2  -D_GNU_SOURCE  -DLIBDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DPREFIX=\"/tmp/deadbeef/deadbeef-devel/_inst\" -DDOCDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/share/doc/\" -DDDB_WARN_DEPRECATED=1 -fvisibility=hidden  ../../../intl/langprefs.c
gcc -c -DLOCALEDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/share/locale\" -DLOCALE_ALIAS_PATH=\"/tmp/deadbeef/deadbeef-devel/_inst/share/locale\" -DLIBDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DBUILDING_LIBINTL -DBUILDING_DLL -DIN_LIBINTL -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DINSTALLDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DNO_XMALLOC -Dset_relocation_prefix=libintl_set_relocation_prefix -Drelocate=libintl_relocate -DDEPENDS_ON_LIBICONV=1 -DHAVE_CONFIG_H -I. -I../../../intl -I..  -g -O2  -D_GNU_SOURCE  -DLIBDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/lib\" -DPREFIX=\"/tmp/deadbeef/deadbeef-devel/_inst\" -DDOCDIR=\"/tmp/deadbeef/deadbeef-devel/_inst/share/doc/\" -DDDB_WARN_DEPRECATED=1 -fvisibility=hidden  ../../../intl/localename.c
In file included from plural.y:35:0:
../../../intl/plural-exp.h:98:23: error: conflicting types for ‘libintl_gettextparse’
 # define PLURAL_PARSE libintl_gettextparse
                       ^
plural.y:40:25: note: in expansion of macro ‘PLURAL_PARSE’
plural.c:187:5: note: in expansion of macro ‘__gettextparse’
../../../intl/plural-exp.h:98:23: note: previous declaration of ‘libintl_gettextparse’ was here
 # define PLURAL_PARSE libintl_gettextparse
                       ^
../../../intl/plural-exp.h:110:12: note: in expansion of macro ‘PLURAL_PARSE’
 extern int PLURAL_PARSE (void *arg);
            ^~~~~~~~~~~~
../../../intl/plural-exp.h:98:23: error: conflicting types for ‘libintl_gettextparse’
 # define PLURAL_PARSE libintl_gettextparse
                       ^
plural.y:40:25: note: in expansion of macro ‘PLURAL_PARSE’
../../../intl/plural.c:63:25: note: in expansion of macro ‘__gettextparse’
 #define yyparse         __gettextparse
                         ^~~~~~~~~~~~~~
plural.c:1111:1: note: in expansion of macro ‘yyparse’
../../../intl/plural-exp.h:98:23: note: previous declaration of ‘libintl_gettextparse’ was here
 # define PLURAL_PARSE libintl_gettextparse
                       ^
../../../intl/plural-exp.h:110:12: note: in expansion of macro ‘PLURAL_PARSE’
 extern int PLURAL_PARSE (void *arg);
            ^~~~~~~~~~~~
plural.c: In function ‘libintl_gettextparse’:
../../../intl/plural.c:64:25: error: too few arguments to function ‘__gettextlex’
 #define yylex           __gettextlex
                         ^
plural.c:1280:16: note: in expansion of macro ‘yylex’
../../../intl/plural.c:64:25: note: declared here
 #define yylex           __gettextlex
                         ^
plural.y:57:12: note: in expansion of macro ‘yylex’
plural.y:156:29: error: ‘arg’ undeclared (first use in this function)
plural.y:156:29: note: each undeclared identifier is reported only once for each function it appears in
make[3]: *** [Makefile:244: plural.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: Leaving directory '/tmp/deadbeef/deadbeef-devel/_build/sub/intl'
make[2]: *** [Makefile:887: all-recursive] Error 1
make[2]: Leaving directory '/tmp/deadbeef/deadbeef-devel/_build/sub'
make[1]: *** [Makefile:623: all] Error 2
make[1]: Leaving directory '/tmp/deadbeef/deadbeef-devel/_build/sub'
make: *** [Makefile:1100: distcheck] Error 1
```
Its fine if you can't review this now, thanks for letting me know! I'll try to see if I can debug the remaining issues on my side in the meantime.

orbea commented:
It seems it was another gettext bug, bumping up the version to `0.19.7` resolved the issue and now it fails the same was as master. I will spend some time seeing what the real minimal version is later.

orbea commented:
The real minimal version seems to be `0.19.2`. So I will update this PR with that change and for reference debian oldstable has `0.19.3`. Also see this link for available gettext versions.

http://ftp.gnu.org/pub/gnu/gettext/

The current version is `0.19.8` which was released in 2016-06-09  and `0.19.2` was released in 2014-07-14. I have no problems in using the most recent version, but using an older version might be preferable for compatibility?

Alexey-Yakovenko commented:
As you've probably noticed from the failed build log on travis-ci, the build machines don't have this version, and there's no easy way to install it without switching to another linux distribution.
Unfortunately, we're not gonna do this any time soon. At least not until 1.8 ships.

orbea commented:
Yes, I see how that could be a problem. What do you mean by 1.8?

Alexey-Yakovenko commented:
The next release of deadbeef is 1.8.

orbea commented:
I think I fixed the per-existing `make distcheck` issue in PR https://github.com/DeaDBeeF-Player/deadbeef/pull/2028.

orbea commented:
I was able to complete `make distcheck` with the older gettext `0.18.2` by using `AM_GNU_GETTEXT([external])`. Online documentation also indicates that this is best way to go, but my changes are based on PR #2028 so I'm not going to push them until that is reviewed.

orbea commented:
@Alexey-Yakovenko Can you reopen this or should I make a new PR? I was going to add another commit that may fix travis, but was waiting for the other PR to be merged which it is now.

Alexey-Yakovenko commented:
reopened

orbea commented:
Thanks! I rebased against master and added one more commit which uses an external gettext which seems to avoid several issues and is also recommended by various online documentation. I can squash the commit if travis passes and you find it agreeable.

Alexey-Yakovenko commented on Makefile.am:
```diff
@@ -48,14 +37,14 @@ deadbeef_SOURCES =\
 	tf.c tf.h\
 	playqueue.c playqueue.h\
 	sort.c sort.h\
+	ConvertUTF/ConvertUTF.c ConvertUTF/ConvertUTF.h\
```
Please change this back. We still prefer to be using libiconv on *nix systems.
ConvertUTF is only for other platforms, where libiconv is not (easily) available.

Alexey-Yakovenko commented on Makefile.am:
```diff
@@ -48,14 +37,14 @@ deadbeef_SOURCES =\
 sdkdir = $(pkgincludedir)
 sdk_HEADERS = deadbeef.h
 
-deadbeef_LDADD = $(LDADD) $(DEPS_LIBS) $(ICONV_LIB) $(DL_LIBS) -lm -lpthread $(INTL_LIBS) plugins/libparser/libparser.a
```
So here you're removing INTL_LIBS from LDADD.. but we still need to link to something to get the gettext APIs. I can't seem to find where / how this is done. Have you checked that localization still works in the builds with your changes?

orbea replied:
Sorry, it seems the last fixup commit broke the translations too. It was working with the previous changes and the newer gettext. I think I am missing something, but adding this back didn't seem to do the trick. I'll try to figure out what I'm missing.

orbea commented:
It seems building, but the translatable strings are not being marked correctly. I think its something missing from the last commit, but I'm not sure what yet.

I'm leaving these links here as they may help explain this.
https://www.gnu.org/software/gettext/manual/html_node/Sources.html#Sources
https://www.gnu.org/software/gettext/manual/html_node/Mark-Keywords.html#Mark-Keywords

Alexey-Yakovenko commented:
> It seems building, but the translatable strings are not being marked correctly. I think its something missing from the last commit, but I'm not sure what yet.

What exactly do you mean by "not being marked correctly"?

Alexey-Yakovenko commented:
> I mean gettext is not seeing the translatable strings and is not doing much of anything at the moment.

Well that was exactly what I was trying to point out in the comment about removing linking to libintl. I think that your changes disabled any localization support which was present in configure.ac

I think that one of the macros which you changed was handling the --enable/disable-nls, and setting the right defined to make it all work. Can you see if those configure options are still present when running `./configure --help`?

Alexey-Yakovenko commented:
you should also check config.h and config.h.in for `ENABLE_NLS`

Alexey-Yakovenko commented:
I tried to figure out why localization doesn't work on you branch, but I couldn't find anything.
It just doesn't work :(

the files get built and installed, ENABLE_NLS=1, LOCALEDIR is correct, nothing seems to be wrong, but refuses to work.

Alexey-Yakovenko commented:
ah! right after writing the last message, I found the wrongness.

`#define LOCALEDIR "/usr/local//locale"` -- this should be /usr/local/share/locale.

It's set from configure.ac like this: ` AC_DEFINE_UNQUOTED(LOCALEDIR, "${prefix}/${DATADIRNAME}/locale", [Name of gettext locale directory])`

So it seems that DATADIRNAME is gone. This variable was set by intltool macros -- most likely IT_PROG_INTLTOOL

Alexey-Yakovenko commented:
It should be safe to set this variable to `share`.

jtojnar commented:
Would not just `${localedir}` be enough? See [Variables for Installation Directories](https://www.gnu.org/software/autoconf/manual/standards.html#Directory-Variables)

Alexey-Yakovenko commented:
I could not find this variable anywhere when I was looking, but if it works -- sure.

Alexey-Yakovenko commented:
@orbea are you still planning to finish this work? Otherwise I'm gonna close the PR.

orbea commented:
Yes, I am interested in seeing this fixed, but I got distracted with many other things and my energy to figure out translations which I don't use is extremely low. I'll close this for now, if someone wants to finish it then they can or I'll come back to it in the future and try to do it myself.
